### PR TITLE
Read-only states for UIA objects

### PR DESCRIPTION
**Pull request by francipvb**
### Link to issue number:
None.

### Summary of the issue:
For various WPF text controls, read-only state isn't added. This is because the current implementation only tries to get the **ValueIsReadOnly** property. However WPF text controls don't support it.

### Description of how this pull request fixes the issue:
This PR adds checks and looks at the document range from the text pattern, if the control implements it. Then it checks for the *Is Read-Only* attribute and adds the state according to the returned value.

### Testing performed:
Tested with output text view of Visual Studio 2019.

### Known issues with pull request:
None known.

### Change log entry:

Section: Changes
* Now NVDA detects more read-only text UIA controls.

### CODE REVIEW

**Comment by francipvb** [comment]
CC @leonardder @josephsl @Adriani90

**Comment by leonardder** [comment]
Thanks for this contribution. I think that the code will be more readable if the STATE_READONLY state is only added once in the code. Something like this:

```
		except COMError:
			isReadOnly=UIAHandler.handler.reservedNotSupportedValue
		if (
			isReadOnly == UIAHandler.handler.reservedNotSupportedValue
			and self.UIATextPattern
		):
			# Most UIA text controls don't support the "ValueIsReadOnly" property, so we need to
			# 	look at root document "IsReadOnly" attribute.
			document = self.UIATextPattern.documentRange
			isReadOnly = document.GetAttributeValue(UIAHandler.UIA_IsReadOnlyAttributeId):
		if isReadOnly != UIAHandler.handler.reservedNotSupportedValue:
			states.add(controlTypes.STATE_READONLY)
```

File: source/NVDAObjects/UIA/__init__.py [hunk 1]
```diff
@@ -1172,12 +1172,24 @@ def _get_states(self):
...
+		if (
+			isReadOnly == UIAHandler.handler.reservedNotSupportedValue
+			and self.UIATextPattern
+		):
+			# Most UIA text controls don't support the "ValueIsReadOnly" property, so we need to look at
+			# 	root document "IsReadOnly" attribute.
+			document = self.UIATextPattern.documentRange
+			isReadOnly = document.GetAttributeValue(UIAHandler.UIA_IsReadOnlyAttributeId)
+		# We can check "isReadOnly" again
+		if isReadOnly:
```
Comment (leonardder) [inline source/NVDAObjects/UIA/__init__.py]: I think reservedNotSupportedValue evaluates to True. I think we either need to do

```suggestion
		if isReadOnly and isReadOnly != UIAHandler.handler.reservedNotSupportedValue:
```

or

```suggestion
		if isReadOnly is True:
```

**Review by LeonarddeR — CHANGES_REQUESTED** [review]

**Review by LeonarddeR — APPROVED** [review]
Thanks for this.

One additional suggestion in which case this new patch improves things.

In the outlook appModule, OutlookUIAWordDocument class, the isReadonlyViewer property can now simply return `controlTypes.STATE_READONLY in self.states`. SO, we no longer have to rely on the object model.

I leave it up to you whether you want to file this as a separate pr after this one, or do it in this one. I think it makes sense to do it as part of this pr.

**Review by LeonarddeR — APPROVED** [review]
Thanks @francipvb, confirmed that Outlook still behaves as expected with this.

**Comment by leonardder** [comment]
@michaelDCurran I think this code looks ok. Could either you or @feerrenrut have a look? This one is pretty small, so shouldn't take longer than a minute or five.

File: source/NVDAObjects/UIA/__init__.py [hunk 2]
```diff
@@ -1231,12 +1231,28 @@ def _get_states(self):
 			states.add(controlTypes.STATE_INVALID_ENTRY)
 		if self._getUIACacheablePropertyValue(UIAHandler.UIA_IsRequiredForFormPropertyId):
 			states.add(controlTypes.STATE_REQUIRED)
+
+		# Read-only state
```
Comment (feerrenrut) [inline source/NVDAObjects/UIA/__init__.py:1235]: Could you please extract this whole section into a new private method `_getIsReadOnlyState()`

**Review by feerrenrut — CHANGES_REQUESTED** [review]
Overall this looks good, thank you @francipvb . Please let me know if you won't have time to extract the method.

**Comment by francipvb** [comment]
Hello,

I don't have time to code on NVDA right now, too much work. I'm sorry...

-- 
Cuando tus fuerzas terminan, las de mi Dios comienzan.

**Review by feerrenrut — APPROVED** [review]

**Pull request merged**